### ISSUE TITLE
Check for compliance: Intro to Python

### ISSUE THREAD

**Issue by leemc-data-ed** [arcus/education_modules#79]
Checking for compliance with QA template 2.0.0

**Comment by leemc-data-ed** [comment]
# Module Quality Assurance Report for PR #[PR number here]
----
Date: {yyyy-mm-dd}
Reviewer: {your name}
qa_template_version: 2.0.0
Name of Module: {take from the title of the main markdown in the PR}
Current Liascript URL: {makes it easy for reviewers and authors to look at content as learners will}
Current Version of Module (use the latest commit value):  {click on the PR and get the clickable short link to the latest commit -- see [quality_assurance_guide.md](quality_assurance_guide.md)}

# Checklist Reports:

## [Directory structure](how_to.md/#Begin-Writing)
* [ ] Folder and file names use lowercase and underscores (no dashes)
* [ ] Main module directory folder name is identical to the name of the module content markdown file.
* [ ] Images, videos, and other audio-visual assets are saved within a `media` folder within the module directory

## Module Organization
* [ ] YAML is at the very top of the file
  * [ ] author name
  * [ ] email
  * [ ] module version number of at least 1.0.0 if first public version or if this is an update then an [appropriately incremented version number](versioning_guidelines.md)
  * [ ] module_template_version number is up to date with the [current sample module](https://raw.githubusercontent.com/arcus/education_modules/main/a_sample_module_template/a_sample_module_template.md) -- if not, the module should be brought in line with any changes that have occurred to the module template before continuing with QA
  * [ ] language
  * [ ] narrator
  * [ ] comment appropriately filled out
  * [ ] long_description appropriately filled out
  * [ ] learning_objectives appropriately filled out
  * [ ] link to CSS (currently https://chop-dbhi-arcus-education-website-assets.s3.amazonaws.com/css/styles.css)
  * [ ] script to Font Awesome kit (currently https://kit.fontawesome.com/83b2343bd4.js)
* [ ] Title is the first line after the end of the YAML
  * [ ] only level-1 header in the entire document.
* [ ] Overview section immediately following Title
  * [ ] surrounded in div with class overview
  * [ ] Comment is at the top of the Overview, linked from YAML rather than rewritten
  * [ ] "Is this module right for me?" contents linked from long_description YAML rather than rewritten
  * [ ] "Estimated time to completion" contents linked from estimated_time YAML rather than rewritten
  * [ ] Prerequisites listed and do not require learner to have specifically attained those skills through any of our other modules.
  * [ ] Learning objectives linked from learning_objectives YAML rather than rewritten
* [ ] All sections following Overview have content (no pages with just header and no additional text / media material).
* [ ] All quizzes start with a level 2 or 3 header. If there is only one quiz in the module, it is labelled "Quiz", or if there are multiple, each header is structured as "Quiz: label" where "label" is a short (~ 1-2 words) description of the content covered in the question(s). E.g., "Quiz: Scatterplots"
* [ ] Educational content ends with a section of additional resources, both ours and outside sources
* [ ] Final section is Feedback.
  * [ ] Learning objectives linked from YAML rather than rewritten
  * [ ] Feedback link is updated appropriately to automatically fill in the module name when clicked by learner.

## Module Content

* [ ] Complexity of the material covered seems appropriate given prerequisites.
* [ ] Time estimate appears accurate for a learner of the targeted level.
* [ ] Clearly defined learning objectives using strong, descriptive verbs. (See [Bloom's taxonomy](https://cft.vanderbilt.edu/guides-sub-pages/blooms-taxonomy/) for ideas.)
* [ ] Every learning objective is covered in the module content.
* [ ] All major topics covered are represented by a learning objective -- there's no significant knowledge imparted that isn't specified in the learning objectives
* [ ] The module title, blurb/comment, and long description accurately reflect the content of the module.
* [ ] Headers are informative and follow a sensible hierarchical structure (the TOC in the left margin should give a good overview of the content covered)
* [ ] Avoids unclear language: unexplained idioms or references, unexplained acronyms, unnecessary technical language.
* [ ] Unusual words, or words taking on a very specific meaning in context, are always defined for the learner, either on the page (e.g. using footnotes) or with links to a definition/glossary.
* [ ] Provides pronunciation guides for any especially unusual words of particular importance (a common example is package names, such as dplyr)
* [ ] Avoids unnecessarily gendered language (e.g. uses "they" singular rather than "he or she" for an unknown person).
* [ ] Informative link text (e.g. instead of "To learn more about python, click [here](www.example.com)", say "Read this article to [learn more about python](www.example.com).")
* [ ] Uses specially formatted highlight boxes consistently and appropriately
* [ ] Short, digestible pieces --- avoids long paragraphs and breaks long sections up with sub-headers
* [ ] All links work (See [Branch References to Change prior to PR Approval](#Branch-References-to-Change-prior-to-PR-Approval) for a place to keep track of internal references that will need to be updated.)
* [ ] Spelling and grammar are correct.


## Code availability
If the module includes code that learners may want to run:
* [ ] A link is provided to run the code in binder
* [ ] A link to the raw code and any other necessary materials (e.g., data) is provided for learners who prefer to work on their own machine

## Formative assessment

* [ ] Sufficient [formative assessment](https://carpentries.github.io/instructor-training/02-practice-learning/#identifying-and-correcting-misconceptions) in the form of quizzes and/or hands-on exercises
* [ ] Clear explanations available after questions unless the nature of the question itself or answer options makes it unnecessary (e.g. a T/F question may not always require follow-up explanation)
* [ ] Quiz questions and hands-on exercises relate directly to learning objectives
* [ ] Every learning objective has at least one associated question
* [ ] Questions can be answered based on the content within the module alone; learners should not need to have read or consulted any of the linked external resources

## Videos and images

* [ ] Screencasts cover a single coherent task so the recording is a short as is feasible. To demonstrate more than one related task, include several short screencasts in succession rather than recording one long screencast.
* [ ] Subtitles available for every recording with audio.
* [ ] Alt text is specified for each image. For help writing and assessing descriptive alt text, see this [alt text image concepts guide](https://www.w3.org/WAI/tutorials/images/) and this [alt text decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/).
  * [ ] Alt text never starts with "photo of", "screenshot of", "image of", or any other information describing the medium of the image (the only exception to this is when describing data visualization -- it is appropriate to start alt text with "a scatterplot...", for example)
  * [ ] Alt text is written to provide the same relevant information to someone using a screen-reader as provided to someone who accesses the information visually
    - E.g., rather than "Screenshot of PR with multiple commits", alt text says 'On a pull request, the PR comment appears at the top. Below, commits appear as their commit summary followed by commit hash.'
  * [ ] For any images that provide a function when clicked, the alt text describes the function of the button rather than describes the visual aspects of the button.
    - E.g., if an image of the GitHub logo is intended to be clicked on to access the education_modules repository, it is labelled "arcus education_modules GitHub repository" rather than "GitHub logo". If this information is also accompanied by a descriptive link in the text, however, then a blank alt tag is appropriate.
  * [ ] Alt text is ideally a maximum of 125 characters, or as close to this as possible.
  * [ ] Purely [decorative visuals](https://www.w3.org/WAI/tutorials/images/decorative/) ([other examples](https://www.portent.com/blog/design-dev/decorative-images-how-to-identify-and-tag-them-for-accessibility.htm)) have a blank alt text label: `![""](image_path)` Note that this is *not* the same as not specifying alt text at all: `![](image_path)`
* [ ] Important visuals (in video, image, or gif) are always described in the audio or in accompanying text.
  - For example, in a screencast, instead of just, "And then click here," provide description that could help scaffold someone without visual access like, "And then click on the button that says 'Run' in the top-right corner of the screen". Be sure to make use of text cues when available (e.g. button labels), not just visual signals like color or location.
  - When important content is conveyed in a visual, describe the key elements. For example, "Running this query produces the table below. It displays the first 5 rows by default, and columns for ID, encounter ID, diagnosis, and outcome."
  - When including a data visualization, describe important features, such as both axis labels and visible trends in the data. For example, "Here's a scatterplot showing number of encounters on the y-axis and age on the x-axis. All 183 patients from our sample are represented here, and it looks like a weak positive trend, with older patients being more likely to have had more encounters. There are a few important outliers, though, such as this patient at about 6 months old with more than 20 encounters already."
  - When visual information is repeated with minimal changes, it's fine to indicate that without providing a full description again. For example, "And here's the updated table, filtered to only show patients who have been seen in the last 2 years."
  - When important visual information in a video is too complex to include sufficient audio description (i.e. it would slow the content down so much as to impair its utility), an alternative video file should be provided with audio descriptions included.
* [ ] Color is never the sole method for distinguishing visual content (including in data visualizations).

## Branch References to Change prior to PR Approval

List here any internal references (stated or hyperlinked) that work now because they refer to the named branch, but will not work once this is on the main branch and the named branch is deleted. Prior to approving the PR and merging to main, all references should be updated in a new commit.

(If there are none, the reviewer can either check off the boxes below without making any edits, or can remove the items below and replace with the text "None".)

* [ ] {description or quote, line ___ in file ____}
* [ ] {description or quote, line ___ in file ____}
* [ ] {description or quote, line ___ in file ____}

## Just Before Approval
* [ ] Once there are no more commits to be made, get the newest commit value for the PR and update the commit value ("Current Version of Module:") at the top of this document.

Congratulations! You can now approve the PR, merge to main, and close (not delete) this issue.

**Comment by leemc-data-ed** [comment]
@drelliche this doesn't have a current PR (because I haven't opened one yet), but I think it's fine to go through the checklist anyway with the current published version? Or maybe it makes more sense for me to get started with what I think needs to be done and open a new PR to trigger the review just like we usually do?

**Comment by cgaynor79** [comment]
@leemc-data-ed Just a small edit in the "Is this module right for me?" section
Python is a versatile programming language that is frequently used for data analysis, machine learning, web development, and more. If _you_ are interested in learning

**Comment by cgaynor79** [comment]
@leemc-data-ed and then another typo/misspell in the prerequisite section: 

 It is helpful to have some _familiarity_

**Comment by cgaynor79** [comment]
@leemc-data-ed in liascript the link to Jupyter binder currently views as html code.
<img width="949" alt="Screen Shot 2022-04-07 at 2 52 29 PM" src="https://user-images.githubusercontent.com/83660822/162276105-7a11de4e-5504-4412-ade3-3781bf341771.png">

**Comment by cgaynor79** [comment]
@leemc-data-ed I just have one piece of feedback that as a beginner to Git, I preferred going to the Jupyter Notebook to work in the cloud because all of the code was already there. I tried the other way too, and I did figure it out, but it actually took me more time than waiting for Jupyter to load! So I think the benefit of the first option is for those that are unfamiliar with using file directories or Github.  And the benefit of the second option is for users that are already familiar or curious to learn more about using Github.I was thinking edits to the first sentences under each option like: "This might work well for you if you either can't or don't want to install Python and Jupyter on your computer and are not familiar with working with directories or Github."  And for the second option: "If you have [Python](https://www.python.org/) and [Jupyter](https://jupyter.org/) installed already on your local computer and have familiarity working with file directories or Github, you might be interested in simply downloading our sample code to your computer."

**Comment by leemc-data-ed** [comment]
@cgaynor79 Thanks for this feedback! I like your wording a lot.

**Comment by cgaynor79** [comment]
@drelliche @leemc-data-ed I worked though the rest of the QA, but I am not sure about the branch reference section? And also not sure about the Directory Structure section - media folder. But from my perspective, I think it's great!

**Comment by leemc-data-ed** [comment]
@cgaynor79 Thanks! As far as the branch reference section, I've made a list of the links that contain references to the `update-intro-to-python` branch, which will need to be changed to reference the main branch once everything else is ready. That'll be one of the last things that will need to happen before we merge.
And regarding the media folder, that item is just making sure that all of the image or video files are in the media folder, which they are (you can see it here: https://github.com/arcus/education_modules/tree/update-intro-to-python/intro_to_python/media).

**Comment by drelliche** [comment]
@leemc-data-ed 

The learning objectives say that learners will "write _text_ and simple code" in the notebook, but I didn't see an opportunity/directions to do that in the notebook. I suggest either taking that out of the learning objective or having a markdown cell early in the notebook with instructions to click on it and write a congratulatory message to yourself for writing text in a Jupyter notebook.

**Comment by cgaynor79** [comment]
@leemc-data-ed  I just checked again, and I am getting a 404 Page Not Found error for the Jupyter Notebook activity. It was working for me the other day...

**Comment by leemc-data-ed** [comment]
@cgaynor79 have you tried opening the notebook in an incognito window? Chrome caching is aggressive, hopefully it should work there though.

**Comment by leemc-data-ed** [comment]
@drelliche good call, I think I'll add that to the notebook! I write about it there but I don't explicitly have them do it, which is an oversight.

**Comment by cgaynor79** [comment]
@leemc-data-ed yep, that worked. The caching is super aggressive. Thanks!

**Comment by leemc-data-ed** [comment]
@drelliche I added a brief Markdown exercise!
@cgaynor79 based on your comments about the exercises in the notebook, I changed the format a little bit so that it's hopefully a little more clear where the learner is expected to do something. I'm curious what you think!

**Comment by leemc-data-ed** [comment]
@cgaynor79 Glad it's clearer, thanks for the feedback!

**Comment by drelliche** [comment]
@leemc-data-ed 

Little notebook edits: 

- After the `.pop` example, "As an aside, the code above contains a comment:" should end with a period, not a colon.
- `list(range(1,20))[7]` example: should the code already be in the cell? In all other cells it is pre-typed
- Final empty cell at the end: remove?

Estimated time / how core is doing the notebook question: Does it make sense to have something in the introduction about how much of the hour is doing hands-on activity in the notebook? And maybe another note on how long it might take when we link to it again? 

Relatedly, the quiz for the hands-on activity seems like it doesn't actually require one to do any of the things in the hands on activity. Here is a structure suggestion (please push back if you don't think it is an improvement)

Jupyter notebooks section:
- current quiz question moved up to after introduction, renamed Quiz: Jupyter notebooks
- Hands-on activity is after that quiz question, nested in the Jupyter notebooks section. 
- Additional quiz question: name three things you learned about in the hands on activity (i.e. lists, dictionaries, loops, strings, whatever) as an open ended question.

**Comment by leemc-data-ed** [comment]
@drelliche This is all great, thanks! Regarding the estimated time for the notebook, @cgaynor79 any idea about how long it took you?

**Comment by cgaynor79** [comment]
@leemc-data-ed I just worked through the hands on activity and it took me just about 30 minutes to complete. I also found a few typos:

First sentence: Jupyter notebooks are a great way _**to**_ use Python
First paragraph, 3rd sentence: The code cells can contain as much or as little code as you like, and the output _**of**_ a cell

Last sentence under the third section: _**mathematical**_ is spelled incorrect

In the explanation about the range function not sure if often frequently makes sense together? : The built-in function range() is _**often frequently**_ used in for loops.  

Also in the above, not sure about the word "for" before loop. You have that a few times in the activity. 

In the It's Your Turn" section the second sentence: They have _**requested**_ that you write some code

That's it from me!

**Comment by pm0kjp** [comment]
The typo thing has gotten me a few times and we should investigate what kind of spell checking is available for Atom, Jupyter (if any), etc.  Just putting this out there as something to think about!

**Comment by leemc-data-ed** [comment]
@cgaynor79 Thanks so much! I'll fix the typos ASAP. Regarding the "for" loop, I've put it in there to distinguish it from a different kind of loop that's called a "while" loop, but I'll try to make that clearer!
@pm0kjp so it looks like it's pretty simple to enable spell checking... but the one that's popular is _just_ a spellchecker (nothing for grammar or anything), so it's not  perfect but it should catch some stuff!

**Comment by leemc-data-ed** [comment]
And I know Atom has some spellchecking as well, though it's not perfect either.

**Comment by leemc-data-ed** [comment]
@cgaynor79 I fixed some of those typos, but others looked to me to be already corrected... I wonder if we're looking at the same version of the notebook (it only really matters for the time estimate, since I've added some extra content in the last few days).

**Comment by cgaynor79** [comment]
@leemc-data-ed Probably a cache clearing thing again. I'm trying it again in an incognito screen. I'll keep you posted.

**Comment by drelliche** [comment]
@pm0kjp @leemc-data-ed @cgaynor79 My version of Atom has a spellcheck package which can be toggled on and off. When on it underlines words it thinks are misspelled, but doesn't do anything for grammar or things like Jupyter/Jupiter.

**Comment by cgaynor79** [comment]
@leemc-data-ed Ok! I got to the updated one in an incognito window...sorry about that mistake! It took me 42 minutes to finish the updated module. I really like the edits and explanations added. I did not completely work through the extra credit add the student identifiers one. I gave it one go, got it wrong, and then looked at the answer. So I think it's safe to say it would to 40-45 minutes.

**Comment by leemc-data-ed** [comment]
Awesome, thanks so much @cgaynor79!

**Comment by drelliche** [comment]
@leemc-data-ed 

Given that it took @cgaynor79 42 minutes to do the notebook and she had already encountered some parts of it, and didn't spend time struggling through the challenge problem, I think the module needs to be split into two if we want to maintain the "maximum 1 hour" length for modules and not leave people thinking it took them so much longer than expected.

Possible split:
**What is Python** ~ 20 m
- The python programming language
- python code
- python scripts
- Jupyter Notebooks
- next steps/downloading python

**Start using python** ~ 1 h
- Hands on Activity

Happy to talk more about all of this!

**Comment by leemc-data-ed** [comment]
@drelliche I was starting to think that as well... In that case I think we could probably call "Intro to Python" pretty much done (I'd have to adjust the learning objectives but that shouldn't take long), but I'd probably want to move a lot of the content of the Jupyter notebook into a new module (which I think would mostly be a lot of copying and pasting). 
One the one hand, that would definitely keep us under our 1 hour time limit... on the other hand, the "what is python" module might seem too easy for most people. What do we think? Any thoughts @pm0kjp?

**Comment by drelliche** [comment]
@leemc-data-ed @pm0kjp 

I think it is reasonable to have a "what is python" module, maybe with a better title, as a survey of ways you can use python. Sure many people will skip over it, and that is fine, but I think it has a lot of value to people who have heard "python, python, python" and really have no idea how to get started.

I'm also not sure that copying the notebook content into the module proper is better than keeping it as is in the module, and just having the module be a framework to get people to the module and check their understanding at the end. I was picturing the "start using python" module has having only a couple of pages in in liascript, mainly a link to the notebook, a quiz, and the feedback page.

**Comment by leemc-data-ed** [comment]
I do like the idea of not having to go back and forth from the module to the notebook (which might happen if we move some of the content out of the notebook)... but I also don't want a situation where it's _necessary_ to have the notebook to get anything useful. A lot will depend on how simple it is to implement the interactive code in liascript (it doesn't seem too hard, but I haven't actually tried it yet). But either way I think I'm inclined to move the actual "writing code in Python" content to its own module. Seem reasonable? Maybe we could follow a similar template to the Intro to R module, add a little more content there, to fill it out a bit?

**Comment by drelliche** [comment]
@leemc-data-ed Sounds very reasonable, and definitely better for the learners! If the Liascript integration is relatively easy, not having a notebook at all might be great (assuming that isn't an absurd amount of work). 

Also, looking through the R intro module, that includes data import so maybe we could combine those modules for python too. I'm also happy to have some shorter modules!

**Comment by leemc-data-ed** [comment]
@franzenr okay I think this is ready for another look!

**Comment by franzenr** [comment]
@leemc-data-ed at this point should I be starting with a new QA comment? (or even a new issue?) It's not clear to me how substantiative any changes have been, and the current checklist has all the boxes ticked (not sure by whom since this has passed many hands at this point)

**Comment by leemc-data-ed** [comment]
@franzenr That makes sense to me, it's changed substantially and really can't be classified as a "compliance check" anymore!

**Comment by franzenr** [comment]
okay, sounds good to me! I'm going to go ahead and close this issue out then. This Phoenix will rise again as QA Intro to Python 2.0.0 🔥 🐦

**Comment by franzenr** [comment]
@leemc-data-ed one related moment of on-the-fly policy setting -- do we think we should mantain the PR or close and open new one?

**Comment by leemc-data-ed** [comment]
This issue has been subsumed by #112
